requests/utils.py

```diff
@@ -615,15 +615,15 @@ def set_environ(env_name, value):
     if value is not None:
         old_value = os.environ.get(env_name)
         os.environ[env_name] = value
-    try:
+        try:
+            yield
+        finally:
+            if old_value is None:
+                del os.environ[env_name]
+            else:
+                os.environ[env_name] = old_value
+    else:
         yield
```

Comment:
@Lukasa refactored the finally statement, let me know if this is more what you would like to see there, also added tests